fix: Crowdin proofreader events are not well triggered - Meeds-io/meeds#2655

sergeByishimo opened the pull request:
The suggestion.approved webhook from Crowdin no longer provides the proofreader information in the payload. Instead, it only includes the translator details.

Now to get proofreader details, we call the Crowdin API after receiving the webhook.

<!-- Ensure to provide github issue and task id in the title -->
<!-- Choose between feat and fix in the title to differenciate a new feature from a fix -->
<!-- Title format must be :
feat: FEATURE TITLE - MEED-XXXX - meeds-io/meeds#1234
or
fix: Fix TITLE - MEED-XXXX - meeds-io/meeds#1234
-->

<!-- Description : describe the feature/the fix by answering theses questions : -->
<!-- Why is this change needed?-->
<!-- Prior to this change, ...-->
<!-- How does it address the issue?-->
<!-- This change ...-->


<!-- Tips : 
Try To Limit Each Line to a Maximum Of 72 Characters
Provide links or keys to any relevant tickets, articles or other resources

Remember to
- Capitalize the subject line
- Use the imperative mood in the subject line
- Do not end the subject line with a period
- Separate subject from body with a blank line
- Use the body to explain what and why vs. how
- Can use multiple lines with "-" for bullet points in body
-->

sergeByishimo committed:
fix: Crowdin proofreader events are not well triggered - Meeds-io/meeds#2655

The suggestion.approved webhook from Crowdin no longer provides the
proofreader information in the payload. Instead, it only includes the
translator details.

Now to get proofreader details, we call the Crowdin API after receiving
the webhook.

AzmiTouil commented on gamification-crowdin-services/src/main/java/io/meeds/crowdin/gamification/utils/Utils.java:
```diff
@@ -34,6 +34,8 @@ public class Utils {
 
   public static final String   WEBHOOKS                           = "/webhooks/";
 
+  public static final String   APPROVALS                           = "/approvals/";
```
```suggestion
  public static final String   APPROVALS                          = "/approvals/";
```

AzmiTouil commented on gamification-crowdin-services/src/main/java/io/meeds/crowdin/gamification/storage/CrowdinConsumerStorage.java:
```diff
@@ -380,6 +377,38 @@ public List<RemoteDirectory> getProjectDirectories(long remoteProjectId,
     }
   }
 
+  public RemoteApproval getApproval(String accessToken, String projectId, String translationId)
+          throws IllegalAccessException {
+
+    try {
+
+      URI uri = URI.create(CROWDIN_API_URL + PROJECTS + projectId + APPROVALS + translationId);
+
+      String response = processGet(uri, accessToken);
+      JSONObject jsonObject = new JSONObject(response).getJSONObject("data");
+
+      RemoteApproval remoteApproval = new RemoteApproval();
+
+      JSONObject userJsonObject = jsonObject.getJSONObject("user");
+
+      remoteApproval.setId(jsonObject.getInt("id"));
+
+      remoteApproval.setUserName(userJsonObject.getString("username"));
+      remoteApproval.setTranslationId(jsonObject.getString("translationId"));
+      remoteApproval.setLanguageId(jsonObject.getString("translationId"));
+      remoteApproval.setStringId(jsonObject.getString("stringId"));
+      remoteApproval.setLanguageId(jsonObject.getString("languageId"));
+
+      return remoteApproval;
+
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException(e);
+    } catch (CrowdinConnectionException e) {
+      throw new IllegalAccessException(TOKEN_EXPIRED_OR_INVALID);
+    }
+  }
```
```suggestion
  public RemoteApproval getApproval(String accessToken, String projectId, String translationId) {
    try {
      URI uri = URI.create(CROWDIN_API_URL + PROJECTS + projectId + APPROVALS + translationId);

      String response = processGet(uri, accessToken);
      JSONObject jsonObject = new JSONObject(response).getJSONObject("data");

      RemoteApproval remoteApproval = new RemoteApproval();
      JSONObject userJsonObject = jsonObject.getJSONObject("user");
      remoteApproval.setId(jsonObject.getInt("id"));
      remoteApproval.setUserName(userJsonObject.getString("username"));
      remoteApproval.setTranslationId(jsonObject.getString("translationId"));
      remoteApproval.setLanguageId(jsonObject.getString("translationId"));
      remoteApproval.setStringId(jsonObject.getString("stringId"));
      remoteApproval.setLanguageId(jsonObject.getString("languageId"));
      return remoteApproval;
    } catch (CrowdinConnectionException e) {
      LOG.warn("Unable to retrieve approval for crowdin translation with id {}.", translationId, e);
      return null;
    }
  }
```

AzmiTouil commented on gamification-crowdin-services/src/main/java/io/meeds/crowdin/gamification/services/WebhookService.java:
```diff
@@ -56,10 +57,22 @@ public List<RemoteProject> getProjects(String accessToken) throws IllegalAccessE
     return crowdinConsumerStorage.getProjects(accessToken);
   }
 
+  public RemoteApproval getApproval(String projectId, String translationId)
+          throws IllegalAccessException, ObjectNotFoundException {
+
+    WebHook webHook = webHookStorage.getWebhookByProjectId(Long.parseLong(projectId));
+
+    if (webHook == null) {
+      throw new ObjectNotFoundException("Webhook with project id '" + projectId + "' doesn't exist");
+    }
+
+    return crowdinConsumerStorage.getApproval(webHook.getToken(), projectId, translationId);
+  }
```
```suggestion
  public RemoteApproval getApproval(String projectId, String translationId) {
    WebHook webHook = webHookStorage.getWebhookByProjectId(Long.parseLong(projectId));
    if (webHook != null) {
       return crowdinConsumerStorage.getApproval(webHook.getToken(), projectId, translationId);
    }
    return null;
  }
```

AzmiTouil commented on gamification-crowdin-services/src/main/java/io/meeds/crowdin/gamification/plugin/SuggestionApprovedTriggerPlugin.java:
```diff
@@ -51,34 +62,42 @@ public List<Event> getEvents(String trigger, Map<String, Object> payload) {
+
+
+    // Retrieve who made that approval by calling Crowdin API
+    try {
+      String translationId = extractSubItem(payload, TRANSLATION, ID);
+
+      RemoteApproval remoteApproval = webhookService.getApproval(getProjectId(payload), translationId);
+
+      if (remoteApproval != null) {
+        eventList.add(new Event(APPROVE_SUGGESTION_EVENT_NAME,
+                remoteApproval.getUserName(),
+                remoteApproval.getUserName(),
+                objectId,
+                TRANSLATION,
+                getProjectId(payload),
+                extractSubItem(payload, TRANSLATION, TARGET_LANGUAGE, ID),
+                extractSubItem(payload, TRANSLATION, PROVIDER) == null,
+                extractSubItem(payload, TRANSLATION, STRING, FILE, DIRECTORY_ID),
+                trigger.equals(SUGGESTION_DISAPPROVED_TRIGGER),
+                countWords(extractSubItem(payload, TRANSLATION, STRING, TEXT))));
+
+      }
+
+    } catch (IllegalAccessException | ObjectNotFoundException e) {
+      LOG.error(e);
```
```suggestion
      String translationId = extractSubItem(payload, TRANSLATION, ID);
      RemoteApproval remoteApproval = webhookService.getApproval(getProjectId(payload), translationId);
      if (remoteApproval != null) {
        eventList.add(new Event(APPROVE_SUGGESTION_EVENT_NAME,
                remoteApproval.getUserName(),
                remoteApproval.getUserName(),
                objectId,
                TRANSLATION,
                getProjectId(payload),
                extractSubItem(payload, TRANSLATION, TARGET_LANGUAGE, ID),
                extractSubItem(payload, TRANSLATION, PROVIDER) == null,
                extractSubItem(payload, TRANSLATION, STRING, FILE, DIRECTORY_ID),
                trigger.equals(SUGGESTION_DISAPPROVED_TRIGGER),
                countWords(extractSubItem(payload, TRANSLATION, STRING, TEXT))));
```

AzmiTouil commented on gamification-crowdin-services/src/main/java/io/meeds/crowdin/gamification/model/RemoteApproval.java:
```diff
@@ -0,0 +1,41 @@
+/*
+ * This file is part of the Meeds project (https://meeds.io/).
+ *
+ * Copyright (C) 2020 - 2024 Meeds Lab contact@meedslab.com
```
```suggestion
 * Copyright (C) 2020 - 2025 Meeds Lab contact@meedslab.com
```

sergeByishimo committed:
fix: Suggested Changes by @AzmiTouil - Meeds-io/meeds#2655

Suggested Changes by @AzmiTouil

sergeByishimo committed:
fix: exception java.lang.IllegalAccessException is never thrown in body of corresponding try statement - Meeds-io/meeds#2655

exception java.lang.IllegalAccessException is never thrown in body of
corresponding try statement

AzmiTouil committed:
Remove useless imports and format code
